### ISSUE TITLE
Add a way to know if a `ComponentElement` is executing its build method or not

### ISSUE THREAD

**Issue by rrousselGit** [flutter/flutter#49505]
## Use case

`provider`'s upcoming feature, `select`, is a way to consume InheritedWidgets that works only inside the `build` method.

Trying to call `select` inside other life-cycles (including `didChangeDependencies`) is not supported.

To make sure that people do not mistakenly still use it in a place where it shouldn't be used, `provider` would like to add an "assert" that fails if the `BuildContext` is not in the progress of building.

## Proposal

This would be implemented through a debug-only flag on `ComponentElement`.
Basically, within `build` that flag would be true, and outside of `build` it wouldn't.

Then, `select` could do:

```dart
extension Select on BuildContext {
  R select<T, R>(R cb(T value)) {
    assert(this is ComponentElement && this.debugIsInsideBuild);
    ...
  }
}
```

**Comment by dnfield** [comment]
I'm pretty sure you can already do this right? `this.owner.debugBuilding` I think is what you want.

**Comment by rrousselGit** [comment]
No, `debugBuilding` is global for the entire tree

This flag would be specific to a given element and is voluntarily false during `initState`/`didChangeDependencies`/...

**Comment by dnfield** [comment]
But once you're `debugBuilding`, all calls to `initState` and `didChangeDependencies` have to be finished anyway for all elements in the tree.

You don't know if you're in the `build` for your specific element, but you know you're after that phase of the lifecycle.

**Comment by dnfield** [comment]
Per the documentation on `buildScope`, which sets `debugBuilding`:

```
  /// Only one [buildScope] can be active at a time.
  ///
  /// A [buildScope] implies a [lockState] scope as well.
```

lockState means no one can call setState.  All state has to be finalized for this cycle by then.

**Comment by rrousselGit** [comment]
> You don't know if you're in the build for your specific element

But that is specifically what I want

> But once you're debugBuilding, all calls to initState and didChangeDependencies have to be finished anyway for all elements in the tree.

That's not true

I already use `debugBuilding` for another assertion, and it may be true during `initState`/...

**Comment by rrousselGit** [comment]
A table of their difference:

| lifecycle | owner.debugBuilding | debugIsInsideBuild
|---|---|---|
| initState | true | false
| didChangeDependencies | true | false
| build | true | true
| event handles (click) | false | false

**Comment by dnfield** [comment]
Ahh ok.  Why do you specifically need to know if you're in build though? Could you explain a bit more about the use case you want to support and what code won't work without this check (or what code is more dangerous without this check)?

**Comment by rrousselGit** [comment]
It is explained in the topmost comment.
It's also showcased in the following test of the branch that implement the said feature: https://github.com/rrousselGit/provider/blob/677173b78dbe9befb331d10b7438c8b04108e5a2/test/context_test.dart#L283

This test showcase how without this flag, the library has no way of preventing users  from calling `select` inside `didChangeDependencies`, which then fails silently.

**Comment by dnfield** [comment]
All the top most comment says is that a new feature called select wants this.  Why?

A quick look at the code doesn't really make it obvious to me .  What about `select` makes it unsafe to call from didChangeDependencies?

**Comment by rrousselGit** [comment]
For more information on the feature in itself, see https://twitter.com/remi_rousselet/status/1218180956139540480?s=19

It's based on a custom Inheritedwidget which works similarly to an InheritedModel.
But the "aspect" is a callback. 

Since it's used almost always with a new closure instance, it means that the InheritedElement needs a way to somehow clear the outdated dependencies when the dependent rebuild.

Otherwise as the dependent keeps rebuilding, it'd end up stacking hundreds of different "Selector", which would all be executed whenever the Inheritedwidget may want to notify dependents. 


I did it by clearing the previous callbacks if `select` is called in a new frame (tracked using addPostFrameCallback).

This works fine. But it means that didChangeDependencies doesn't work with this `select` function.

It is not a problem (as the closure act as a didChangeDependencies if needed), but may be unexpected to some users.

**Comment by rrousselGit** [comment]
This could be summarised as "to work, `select` relies on a workaround to #12992 which makes the compatible incompatible with `didChangeDependencies`".

To begin with, the reason #12992 can't be fixed easily is because of `didChangeDependencies`.
This feature purposefully gave up on that life-cycle to work.

**Comment by dnfield** [comment]
It sounds like you're working around an inability to clear out stale dependencies - wouldn't it be better to just have a way to do that without postframecallbacks or checking whether your dependent is calling you from build or not?

**Comment by rrousselGit** [comment]
Yes, supporting didChangeDependencies and removing the need for post frame callbacks is definitely better

That would imply a much bigger maintainability cost on the framework side though. 
Like being able to know at runtime in release if an Element is within build or didChangeDependencies or outside both

If the team is fine with fully tackling such thing, I'm of course fine with it. 

But this it's only for a library, and being able to use `select` in didChangeDependencies is not critical at all.
So I'm not sure it's worth the effort, compared to a simple debug flag

**Comment by rrousselGit** [comment]
Similarly, #12992 or not, didChangeDependencies would still be an issue with `select` because of how that lifecycle currently works.

If multiple Inheritedwidgets updates within the same frame, the lifecycle will be called multiple times. 
This in return will create multiple duplicate selectors, which is undesired. 

And I'm still convinced that didChangeDependencies is unsafe (https://github.com/flutter/flutter/issues/34130) and that in all situations we can just use the build method instead.

So not supporting didChangeDependencies is not really a problem in my opinion.

**Comment by dnfield** [comment]
So I have two thoughts here.  

1. On the topic of this specific issue: exposing new API surface is expensive. It limits future decisions, it increases the likelyhood that some change we'll want to make later is breaking, and it may be confusing to users who think it means something it doesn't.

1. The proposed reason to support this API seems dangerous to me. `select` looks like a cheap method on `BuildContext` to only rebuild when certain properties on an object you care about change, but it actually ends up adding more work each and every frame per property you select (?!), even if you use it correctly so that it is only called within build methods.

I think the idea behind `select` is a really good one, but I'm wondering if there's some alternative implementation of it that does not hide so much work for the sake of simpler code...

**Comment by rrousselGit** [comment]
> The proposed reason to support this API seems dangerous to me. select looks like a cheap method on BuildContext to only rebuild when certain properties on an object you care about change, but it actually ends up adding more work each and every frame per property you select (?!), even if you use it correctly so that it is only called within build methods.


It's not every frame.
It's only on `InheritedElement.notifyClient`.

**Comment by dnfield** [comment]
Ok - but that could get called every frame if you decide to update your model every frame for some reason, such as having it tied to some animated value.

Basically, it makes it very easy to end up doing something that's very expensive - and it seems like it could be rewritten in such a way where it could still deliver a lot of value but not make the very expensive thing so easy to do.

**Comment by rrousselGit** [comment]
I don't see how that is more expensive than what InheritedModel does in general

It is almost identical to:

```dart
class MyModel extends InheritedModel<bool Function(MyModel m)> {
  @override
  bool updateShouldNotify(MyModel oldWidget) => true;

  @override
  bool updateShouldNotifyDependent(
    InheritedModel<_SelectorAspect<MyModel>> oldWidget,
    Set<_SelectorAspect<MyModel>> dependencies,
  ) {
    return dependencies.any((cb) => cb(this));
  }
}
```

**Comment by rrousselGit** [comment]
Ironically, the changes in https://github.com/flutter/flutter/pull/49527 kind of put this debug flag in an awkward position.

It wouldn't be possible to implement such a flag directly from `ComponentElement` anymore if `State.didChangeDependencies` is called from inside `ComponentElement.build`.

I suppose I can live without such a flag as I have a few other asserts that should somehow trigger during development, just slightly less convenient.
So this can probably be closed if we want to go for https://github.com/flutter/flutter/pull/49527

I'm still curious on how you'd implement such feature @dnfield though 😃

**Comment by dnfield** [comment]
This is part of why it's not always ideal to expose flags like this - the change in #49527 is something that should help runtime performance.

As far as how to implement this - I would avoid encouraging people registering multiple callbacks Each `select` call right now triggers a whole bunch of notifications - instead it could either track state for the user or require that it only be called once with all of the relevant properties you want to subscribe to.

**Comment by rrousselGit** [comment]
The documentation is enough to discourage people using many selectors at once.

But having the ability to have multiple ones is critical and there's no reason to remove it.

Its better not to have such debug flag than to remove the ability to call `select` multiple times.

**Comment by escamoteur** [comment]
Could other users besides provider benefit from this change?

**Comment by rrousselGit** [comment]
`flutter_hook` needed the exact same flag too. Although in that very specific case, I was able to implement it myself.

**Comment by rrousselGit** [comment]
Although now that I think about it, the changes in #49527 means flutter_hooks will need this flag too, as the one I hacked myself will no longer work.
